### PR TITLE
M3-1694 Sidebar now displays regardless of user level...

### PR DESCRIPTION
**Pull request by rjstires**
## Notes
This issue was actually a result of a condition checking 'managed', not the settings themselves. 

- Split PrimaryNav into smaller components.
- Added user grants request to profile request flow.
- Used grants to hide/show account options in navigation bar.,
- Corrected issue with domain record types.

## Changelog
change: PrimaryNav broken down into smaller components.
~add: AccountNavigation component (extracted from PrimaryNav)~
add: ThemeToggle component (extracted from PrimaryNav)
add: Service to get the current user's grants.
add: The current user's grants are now requested and added to the Redux store at `__resoureces.profile.grants`.

## Type Changes
change: Profile type now includes grants.
change: Record renamed DomainRecord to prevent collision with native Record type.
change: Used native Record type for Grants.
add: GrantTypes added to provide key types for Grants.

## Updates
add: ApplicationIsOnFire catch all component. With any luck, no one will ever see this.
change: Added componentDidCatch LCM to App for catch all display of ApplicationIsOnFire.
change: In the event of a profile request error, display ApplicationIsOnFire component.

### CODE REVIEW

**Comment by patthiel** [comment]
In testing this, it seems the unrestricted user can no longer view the "Account" side bar menu item

File: src/components/PrimaryNav/PrimaryNav.tsx [hunk 1]
```diff
@@ -294,130 +258,47 @@ class PrimaryNav extends React.Component<CombinedProps, State> {
...
+            [classes.fadeContainer]: true,
+          })}>
+
+            {primaryLinks.map(primaryLink => this.renderPrimaryLink(primaryLink))}
+
+            <AccountNavigation
+              closeMenu={closeMenu}
+              expandedMenus={expandedMenus}
+              linkIsActive={this.linkIsActive}
+              expandMenutItem={this.expandMenutItem}
```
Comment (patthiel) [inline src/components/PrimaryNav/PrimaryNav.tsx]: If i'm reading this correctly, this might be missing `hasAccountAccess`

Comment (rjstires) [reply src/components/PrimaryNav/PrimaryNav.tsx]: You're reading incorrectly, `hasAccountAccess` is provided by Redux.

File: src/store/reducers/resources/profile.ts [hunk 2]
```diff
@@ -57,11 +57,22 @@ export default (state: State = DEFAULT_STATE, action: Action) => {
...
 
 export const requestProfile = () => (dispatch: Dispatch<State>) => {
 
   dispatch(startRequest());
   getProfile()
-    .then(compose(dispatch, handleSuccess, prop('data')))
+    .then(response => response.data)
+    .then(maybeRequestGrants)
+    .then(compose(dispatch, handleSuccess))
     .catch(compose(dispatch, handleError));
```
Comment (Jskobos) [inline src/store/reducers/resources/profile.ts]: This error handling can be tweaked, e.g. may want to put the async method in a `try` block. Mocking or blocking `getMyGrants` crashes the app. Also the handleError function is getting called with some weird stuff when I block the request.

**Comment by rjstires** [comment]
Thank you @patthiel and @Jskobos for the review and feedback. Its updated, rebased, and ready!

**Comment by rjstires** [comment]
Rebased again.

File: src/features/TheApplicationIsOnFire.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+});
+
+interface Props { }
+
+type CombinedProps = Props & WithStyles<ClassNames>;
+
+const TheApplicationIsOnFire: React.StatelessComponent<CombinedProps> = (props) => {
+  return (
+    <Dialog open>
+      <DialogTitle>Opps!</DialogTitle>
```
Comment (Jskobos) [inline src/features/TheApplicationIsOnFire.tsx]: Oops

Comment (rjstires) [reply src/features/TheApplicationIsOnFire.tsx]: Opp!

File: src/components/PrimaryNav/PrimaryNav.tsx [hunk 4]
```diff
@@ -16,8 +15,9 @@ import isPathOneOf from 'src/utilities/routing/isPathOneOf';
 import ThemeToggle from './ThemeToggle';
 
 interface PrimaryLink {
-  display: string,
-  href: string,
+  display: string;
+  href: string;
+  key: string;
```
Comment (rjstires) [inline src/components/PrimaryNav/PrimaryNav.tsx]: So I can add the data-qa attribute.

File: src/components/PrimaryNav/PrimaryNav.tsx [hunk 5]
```diff
@@ -158,7 +158,7 @@ interface State {
 
 type CombinedProps = Props & StateProps;
 
-class PrimaryNav extends React.Component<CombinedProps, State> {
+export class PrimaryNav extends React.Component<CombinedProps, State> {
```
Comment (rjstires) [inline src/components/PrimaryNav/PrimaryNav.tsx]: So I can test it!

File: src/components/PrimaryNav/PrimaryNav.tsx [hunk 6]
```diff
@@ -186,27 +186,55 @@ class PrimaryNav extends React.Component<CombinedProps, State> {
   }
 
   createMenuItems = () => {
-    const { hasAccountAccess } = this.props;
+    const {
+      hasAccountAccess,
+      // isLongviewEnabled,
```
Comment (rjstires) [inline src/components/PrimaryNav/PrimaryNav.tsx]: All of these commented bits are for when we eventually are told to hide LV unless they already have it.

File: src/components/PrimaryNav/PrimaryNav.tsx [hunk 7]
```diff
@@ -186,27 +186,55 @@ class PrimaryNav extends React.Component<CombinedProps, State> {
...
+      isManagedAccount,
+    } = this.props;
 
     const primaryLinks = [
```
Comment (rjstires) [inline src/components/PrimaryNav/PrimaryNav.tsx]: The intent here is kind of obvious, we'll be able to extract from the grants whether or not we even show the menu items. So here's the scaffolding.

File: src/components/PrimaryNav/PrimaryNav.tsx [hunk 8]
```diff
@@ -186,27 +186,55 @@ class PrimaryNav extends React.Component<CombinedProps, State> {
...
+
     if (hasAccountAccess) {
-      primaryLinks.push({ display: 'Account', href: '/account' });
+      primaryLinks.push({ display: 'Account', href: '/account', key: 'account' });
     }
 
-    this.setState({
-      primaryLinks,
-    })
+    primaryLinks.push({ display: 'Get Help', href: '/support', key: 'support' });
```
Comment (rjstires) [inline src/components/PrimaryNav/PrimaryNav.tsx]: I seen "Get Help" was it's own jawn, and it seemed to be that way so we didn't display the divider. I've baked in logic to not display the divider if it's the last item in the array.

File: src/components/PrimaryNav/PrimaryNav.tsx [hunk 9]
```diff
@@ -258,6 +286,7 @@ class PrimaryNav extends React.Component<CombinedProps, State> {
           ${classes.listItem}
           ${this.linkIsActive(primaryLink.href) && classes.active}
         `}
+        data-qa-nav-item={primaryLink.key}
```
Comment (rjstires) [inline src/components/PrimaryNav/PrimaryNav.tsx]: Because test.

File: src/components/PrimaryNav/PrimaryNav.tsx [hunk 10]
```diff
@@ -300,31 +329,8 @@ class PrimaryNav extends React.Component<CombinedProps, State> {
             [classes.fadeContainer]: true,
           })}>
 
-            {this.state.primaryLinks.map(primaryLink => this.renderPrimaryLink(primaryLink))}
-
-            <ListItem
```
Comment (rjstires) [inline src/components/PrimaryNav/PrimaryNav.tsx]: Moved into the PrimaryLinks array.

File: src/components/PrimaryNav/PrimaryNav.tsx [hunk 11]
```diff
@@ -386,39 +392,47 @@ class PrimaryNav extends React.Component<CombinedProps, State> {
...
+    return true;
+  }
 
-  /** Data hasnt loaded yes, lock it down. */
-  if (!profile) {
-    return {
-      hasAccountAccess: false,
-    }
+  const { grants } = profile;
+  if (!grants) {
```
Comment (rjstires) [inline src/components/PrimaryNav/PrimaryNav.tsx]: I dont think this is technically possible, but Im sure one extra condition wont bring the software to a fire death.

File: src/components/PrimaryNav/PrimaryNav.tsx [hunk 12]
```diff
@@ -386,39 +392,47 @@ class PrimaryNav extends React.Component<CombinedProps, State> {
...
 const connected = connect(mapStateToProps);
 
-const enhanced: any = compose(
-  withStyles(styles),
-  withRouter,
-  connected,
-);
-
-export default enhanced(PrimaryNav);
+export default withStyles(styles)(withRouter(connected(PrimaryNav)));
```
Comment (rjstires) [inline src/components/PrimaryNav/PrimaryNav.tsx]: Ramda's type inference on compose React components is teh sux.

**Pull request merged**